ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/PullQueryStreamWriter.java

```diff
@@ -0,0 +1,215 @@
+import java.time.Clock;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings("checkstyle:CyclomaticComplexity")
+public class PullQueryStreamWriter implements StreamingOutput {
+  private static final Logger LOG = LoggerFactory.getLogger(PullQueryStreamWriter.class);
+
+  private static final int FLUSH_SIZE_BYTES = 50 * 1024;
+  private static final long MAX_FLUSH_MS = 30;
+
+  private final long disconnectCheckInterval;
+  private final PullQueryQueue pullQueryQueue;
+  private final Clock clock;
+  private final PullQueryResult result;
+  private final ObjectMapper objectMapper;
+  private volatile boolean limitReached = false;
+  private volatile boolean connectionClosed;
+  private volatile Throwable pullQueryException;
+  private boolean closed;
+  private boolean sentAtLeastOneRow = false;
+
+  PullQueryStreamWriter(
+      final PullQueryResult result,
+      final long disconnectCheckInterval,
+      final ObjectMapper objectMapper,
+      final PullQueryQueue pullQueryQueue,
+      final Clock clock,
+      final CompletableFuture<Void> connectionClosedFuture
+  ) {
+    this.result = Objects.requireNonNull(result, "result");
+    this.objectMapper = Objects.requireNonNull(objectMapper, "objectMapper");
+    this.disconnectCheckInterval = disconnectCheckInterval;
+    this.pullQueryQueue = Objects.requireNonNull(pullQueryQueue, "pullQueryQueue");
+    this.clock = Objects.requireNonNull(clock, "clock");
+    connectionClosedFuture.thenAccept(v -> connectionClosed = true);
+    result.onException(t -> pullQueryException = t);
+    result.onCompletion(v -> limitReached = true);
+  }
+
+  @Override
+  public void write(final OutputStream output) {
+    try {
+      StringBuilder sb = new StringBuilder();
+      PullQueryRow head = null;
+      long lastFlush = clock.millis();
+
+      final StreamedRow header
+          = StreamedRow.header(result.getQueryId(), result.getSchema());
+      sb.append("[").append(writeValueAsString(header));
+
+      while (!connectionClosed && !pullQueryQueue.isClosed() && !limitReached) {
+
+        final PullQueryRow row = pullQueryQueue.pollRow(
+            Math.min(disconnectCheckInterval, MAX_FLUSH_MS),
+            TimeUnit.MILLISECONDS
+        );
+        if (row != null) {
+          // We keep track of a head of the queue so that we know when writing the current row if
+          // there a next.  This is to write proper json
+          final PullQueryRow toProcess = head;
+          head = row;
+          if (toProcess == null) {
+            continue;
+          }
+          writeRow(toProcess, head, sb);
+          if (sb.length() >= FLUSH_SIZE_BYTES || (clock.millis() - lastFlush) >= MAX_FLUSH_MS) {
+            output.write(sb.toString().getBytes(StandardCharsets.UTF_8));
+            output.flush();
```

Comment:
I'm wondering how this works from a client perspective. It looks like we'd be flushing incomplete JSON, right? Is that what was happening previously for pull queries? Would clients know how to handle that properly?

Reply:
For my own clarification: were you referring to the JSON format of the HTTP response, or the JSON format of a single record?